Userland/Applications/PixelPaint/ImageMasking.cpp

```diff
@@ -0,0 +1,181 @@
+        done(ExecResult::Cancel);
+    };
+
+    generate_new_mask();
+}
+
+void ImageMasking::revert_possible_changes()
+{
+    if (m_did_change && m_reference_mask) {
+        MUST(m_editor->active_layer()->set_bitmaps(m_editor->active_layer()->content_bitmap(), m_reference_mask.release_nonnull()));
+        m_editor->layers_did_change();
+    }
+    cleanup_resources();
+}
+
+void ImageMasking::generate_new_mask()
+{
+    (void)ensure_reference_mask();
+    if (m_reference_mask.is_null())
+        return;
+
+    int min_luminosity_start = m_edge_masking_slider->lower_range();
+    int min_luminosity_full = m_full_masking_slider->lower_range();
+    int max_luminosity_full = m_full_masking_slider->upper_range();
+    int max_luminosity_end = m_edge_masking_slider->upper_range();
+    int current_content_luminosity, approximation_alpha;
+    bool has_start_range = min_luminosity_start != min_luminosity_full;
+    bool has_end_range = max_luminosity_end != max_luminosity_full;
+    Gfx::Color reference_mask_pixel, content_pixel;
+
+    for (int y = 0; y < m_reference_mask->height(); y++) {
+        for (int x = 0; x < m_reference_mask->width(); x++) {
+            reference_mask_pixel = m_reference_mask->get_pixel(x, y);
+            if (!reference_mask_pixel.alpha())
+                continue;
+
+            content_pixel = m_editor->active_layer()->content_bitmap().get_pixel(x, y);
+            if (!content_pixel.alpha()) {
+                m_editor->active_layer()->mask_bitmap()->set_pixel(x, y, reference_mask_pixel.with_alpha(0));
+                continue;
+            }
+
+            current_content_luminosity = content_pixel.luminosity();
+
+            if (current_content_luminosity >= min_luminosity_start && current_content_luminosity < min_luminosity_full && has_start_range) {
+                approximation_alpha = reference_mask_pixel.alpha() * static_cast<float>((current_content_luminosity - min_luminosity_start)) / (min_luminosity_full - min_luminosity_start);
+                m_editor->active_layer()->mask_bitmap()->set_pixel(x, y, reference_mask_pixel.with_alpha(approximation_alpha));
+            } else if (current_content_luminosity > max_luminosity_full && current_content_luminosity <= max_luminosity_end && has_end_range) {
+                approximation_alpha = reference_mask_pixel.alpha() * (1 - static_cast<float>((current_content_luminosity - max_luminosity_full)) / (max_luminosity_end - max_luminosity_full));
+                m_editor->active_layer()->mask_bitmap()->set_pixel(x, y, reference_mask_pixel.with_alpha(approximation_alpha));
+            } else if (current_content_luminosity >= min_luminosity_full && current_content_luminosity <= max_luminosity_full) {
+                m_editor->active_layer()->mask_bitmap()->set_pixel(x, y, reference_mask_pixel);
+            } else {
+                m_editor->active_layer()->mask_bitmap()->set_pixel(x, y, reference_mask_pixel.with_alpha(0));
```

Comment:
Let's apply a bit of DRY here and not repeat `m_editor->active_layer()->mask_bitmap()->set_pixel(x, y, ...` 5 times - it's only the pixel color that needs to be determined.